File: algorithms/visualstudio/algorithms.csharp/DivideAndConquer/MergeSortCountInversions.cs
using System.Collections.Generic;
using System.Linq;
using System.Numerics;

namespace algorithms.csharp.DivideAndConquer
{
    public class ListAndInversions
    {
        public List<int> List { get; private set; }
        public BigInteger CountOfInversions { get; private set; }

        public ListAndInversions(List<int> list, int countOfInversions = 0) : this(list, new BigInteger(countOfInversions)) { }
        
        public ListAndInversions(List<int> list, BigInteger countOfInversions)
        {
            List = list;
            CountOfInversions = countOfInversions;
        }
    }

    public class MergeSortCountInversions
    {
        public BigInteger CountInversions(List<int> listToCountInversionsIn)
        {
            return CountInversions(new ListAndInversions(listToCountInversionsIn)).CountOfInversions;
        }

        private ListAndInversions CountInversions(ListAndInversions listAndInversions)
        {
            if (listAndInversions.List.Count < 2)
            {
                return listAndInversions;
            }

            var splitPoint = listAndInversions.List.Count / 2;

            var list1 = listAndInversions.List.Take(splitPoint).ToList();
            var list2 = listAndInversions.List.Skip(splitPoint).ToList();

            var listAndInversions1 = new ListAndInversions(list1);
            var listAndInversions2 = new ListAndInversions(list2);

            var sortedListAndInversions1 = CountInversions(listAndInversions1);
            var sortedListAndInversions2 = CountInversions(listAndInversions2);

            return MergeAndCountInversions(sortedListAndInversions1, sortedListAndInversions2);
        }
        
        public ListAndInversions MergeAndCountInversions(ListAndInversions listAndInversions1,
            ListAndInversions listAndInversions2)
        {
            var mergedList = new List<int>();
            var list1 = listAndInversions1.List;
            var list2 = listAndInversions2.List;

            var currentInversions = listAndInversions1.CountOfInversions + listAndInversions2.CountOfInversions;

            var list1Index = 0;
            var list2Index = 0;
            
            while (list1Index < list1.Count && list2Index < list2.Count)
            {
                var currentList1Element = list1.ElementAt(list1Index);
                var currentList2Element = list2.ElementAt(list2Index);

                if (currentList1Element <= currentList2Element)
                {
                    mergedList.Add(currentList1Element);
                    list1Index++;
                }
                else
                {
                    var numberOfInversionsToAdd = list1.Count - list1Index;
                    currentInversions = currentInversions + numberOfInversionsToAdd;
                    mergedList.Add(currentList2Element);
                    list2Index++;
                }
            }

            if (list1.Count == list1Index)
            {
                var tailOfList2 = list2.Skip(list2Index);
                mergedList.AddRange(tailOfList2);
            }

            if (list2.Count == list2Index)
            {
                var tailOfList1 = list1.Skip(list1Index);
                mergedList.AddRange(tailOfList1);
            }

            return new ListAndInversions(mergedList, currentInversions);
        }
    }
}

File: algorithms/visualstudio/algorithms.fsharp/MergeSortCountInversions.fs
namespace algorithms.fSharp.divideAndConquer

module MergeSortCountInversions =
    
    type ListAndInversions (list: List<int>, ?numberOfInversions) =
        member this.list = list
        member this.numberOfInversions = defaultArg numberOfInversions 0
                

    let mergeAndCountInversions (listAndInversionsLeft: ListAndInversions, listAndInversionsRight: ListAndInversions) =    
             
        let rec mergeAndCountInversionsPrivate (left: ListAndInversions, right: ListAndInversions, result: ListAndInversions) =      
            match (left.list, right.list) with
            | ([],_) -> ListAndInversions(List.append result.list right.list, result.numberOfInversions)
            | (_,[]) -> ListAndInversions(List.append result.list left.list, result.numberOfInversions)
            | (x::xs,y::ys) when y >= x  ->  mergeAndCountInversionsPrivateRec (x, xs, right.list, result, 0)
            | (x::xs,y::ys) when x > y   ->  mergeAndCountInversionsPrivateRec (y, left.list, ys, result, left.list.Length)
        
        and mergeAndCountInversionsPrivateRec (nextNumberToAddToOutputList: int, newLeftList: List<int>, newRightList: List<int>, oldResult: ListAndInversions, numberToAddToInversions: int) =
            let newResultList = List.append oldResult.list [nextNumberToAddToOutputList]
            let newResultInversions = numberToAddToInversions + oldResult.numberOfInversions

            mergeAndCountInversionsPrivate(ListAndInversions(newLeftList), ListAndInversions(newRightList), ListAndInversions(newResultList, newResultInversions))

        let countOfInversions = listAndInversionsLeft.numberOfInversions + listAndInversionsRight.numberOfInversions
        mergeAndCountInversionsPrivate (listAndInversionsLeft, listAndInversionsRight, ListAndInversions([], countOfInversions))


    let countInversions list = 
        let rec countInversionsRec (listAndInversions: ListAndInversions) =
            match listAndInversions with
            | listAndInversions when listAndInversions.list.Length < 2 -> listAndInversions
            | _ -> countInversionsRecAnd listAndInversions

        and countInversionsRecAnd listAndInversions = 
            let splitPoint = listAndInversions.list.Length / 2
            let list1 = Seq.toList <| (Seq.take splitPoint <| listAndInversions.list)
            let list2 = Seq.toList <| (Seq.skip splitPoint <| listAndInversions.list)

            let listAndInversions1 = ListAndInversions(list1)
            let listAndInversions2 = ListAndInversions(list2)

            let sortedListAndInversions1 = countInversionsRec listAndInversions1
            let sortedListAndInversions2 = countInversionsRec listAndInversions2

            mergeAndCountInversions (sortedListAndInversions1, sortedListAndInversions2)


        countInversionsRec (ListAndInversions(list))
        |> fun elem -> elem.numberOfInversions